### PR TITLE
Docker info without VCH info or container count

### PR DESCRIPTION
**Pull request by sflxn**
This is a WIP for Docker Info.  It is still waiting on container metadata in the portlayer.  It also needs portlayer support to get resource pool info.

### CODE REVIEW

File: lib/apiservers/engine/backends/system.go [hunk 1]
```diff
@@ -15,31 +15,148 @@
...
 }
 
+type ContainerStatus struct {
+	count      int
+	numRunning int
+	numStopped int
+	numPaused  int
+}
+
+var (
```
Comment (caglar10ur) [inline lib/apiservers/engine/backends/system.go]: const?

File: lib/apiservers/engine/backends/system.go [hunk 2]
```diff
@@ -15,31 +15,148 @@
...
+
 func (s *System) SystemInfo() (*types.Info, error) {
-	Driver := "Portlayer Storage"
-	IndexServerAddress := "https://index.docker.io/v1/"
-	ServerVersion := "0.0.1"
-	Name := "VIC"
+	// Set VIC specific data
+	Driver := s.ProductName + " Backend Engine"
+	IndexServerAddress := "registry-1.docker.io"
+	ServerVersion := "0.3"
```
Comment (caglar10ur) [inline lib/apiservers/engine/backends/system.go]: Same for my comment on the other PR, we should make this a file or a build tag

File: lib/apiservers/engine/backends/system.go [hunk 3]
```diff
@@ -15,31 +15,148 @@
...
+)
+
 func (s *System) SystemInfo() (*types.Info, error) {
-	Driver := "Portlayer Storage"
-	IndexServerAddress := "https://index.docker.io/v1/"
-	ServerVersion := "0.0.1"
-	Name := "VIC"
+	// Set VIC specific data
+	Driver := s.ProductName + " Backend Engine"
+	IndexServerAddress := "registry-1.docker.io"
```
Comment (caglar10ur) [inline lib/apiservers/engine/backends/system.go]: make this const

File: lib/apiservers/engine/backends/system.go [hunk 4]
```diff
@@ -15,31 +15,148 @@
...
+		ContainersRunning:  containerStatus.numRunning,
+		ContainersPaused:   containerStatus.numPaused,
+		ContainersStopped:  containerStatus.numStopped,
+		Images:             numImages,
+		SystemStatus:       systemStatus,
+		Debug:              (os.Getenv("DEBUG") != ""),
+		NGoroutines:        runtime.NumGoroutine(),
+		SystemTime:         time.Now().Format(time.RFC3339Nano),
+		ExecutionDriver:    Driver,
+		LoggingDriver:      "json-file",
```
Comment (caglar10ur) [inline lib/apiservers/engine/backends/system.go]: What is json-file?

File: lib/apiservers/engine/backends/system.go [hunk 5]
```diff
@@ -15,31 +15,148 @@
...
+		CgroupDriver:       "N/A",
+		// FIXME: Get this info once we have event listening service
+		//	NEventsListener    int
+		DockerRootDir:    "N/A",
+		ClusterStore:     "",
+		ClusterAdvertise: "",
+		HTTPProxy:        sockets.GetProxyEnv("http_proxy"),
+		HTTPSProxy:       sockets.GetProxyEnv("https_proxy"),
+		NoProxy:          sockets.GetProxyEnv("no_proxy"),
```
Comment (caglar10ur) [inline lib/apiservers/engine/backends/system.go]: We don't need a package for this, just call os.GetEnv :)

File: isos/base.sh [hunk 6]
```diff
@@ -68,7 +68,7 @@ cp $DIR/base/*.repo $(rootfs_dir $PKGDIR)/etc/yum.repos.d/
 cp $DIR/base/yum.conf $(rootfs_dir $PKGDIR)/etc/yum/
 
 # install the core packages
-yum_cached -c $cache -u -p $PKGDIR install filesystem coreutils bash linux-esx --nogpgcheck -y
+yum_cached -c $cache -u -p $PKGDIR install filesystem coreutils bash linux-esx photon-release --nogpgcheck -y
```
Comment (hickeng) [inline isos/base.sh]: This should be added into the appliance-staging.sh, not the base. The entire initramfs is blown away after tether launches in the bootstrap.

File: lib/apiservers/engine/backends/system.go [hunk 7]
```diff
@@ -15,31 +15,148 @@
...
+	systemStatus := make([][2]string, 1)
+	systemStatus[0][0] = s.ProductName + " Backend Engine"
+	if pingPortlayer(plClient) {
+		systemStatus[0][1] = "RUNNING"
+	} else {
+		systemStatus[0][1] = "STOPPED"
+	}
+
+	// Retrieve info about the VCH from Exec port layer
+	//FIXME: Write the portlayer code to do this
```
Comment (hickeng) [inline lib/apiservers/engine/backends/system.go]: This doesn't need portlayer code, it's part of the tether/extraconfig. You can reload a portion of extraconfig and get the current status for all components.

File: lib/apiservers/engine/backends/system.go [hunk 8]
```diff
@@ -15,31 +15,148 @@
...
+		//	MemoryLimit        bool
+		//	SwapLimit          bool
+		KernelMemory: false,
+		//	IPv4Forwarding     bool
+		//	BridgeNfIptables   bool
+		//	BridgeNfIP6tables  bool `json:"BridgeNfIp6tables"`
+	}
+
+	info.Plugins.Network = append(info.Plugins.Network, "N/A")
+	info.Plugins.Volume = append(info.Plugins.Volume, "N/A")
```
Comment (hickeng) [inline lib/apiservers/engine/backends/system.go]: We should use this to identify the possible `locations` that can be optionally specified for volume create and eventually the free space. These are the map keys in VirtualContainerHostConfigSpec.VolumeLocations.

File: lib/apiservers/engine/backends/system.go (hunk 5 above)
Comment (hickeng) [inline lib/apiservers/engine/backends/system.go]: Why do this via the environment at all - it's not likely to be set in the appliance.
If we need a proxy then it should be in the VCH config struct so that imagec can make use of it as well.

File: lib/apiservers/engine/backends/system.go (hunk 4 above)
Comment (hickeng) [inline lib/apiservers/engine/backends/system.go]: Suggest we change this - logging driver doesn't really exist for us but is one of the first places where we should consider extensibility - being able to log to syslogd (which allows LogInsight) is very useful enterprise function.

File: lib/apiservers/engine/backends/system.go (hunk 1 above)
Comment (sflxn) [inline lib/apiservers/engine/backends/system.go]: Good point.

File: lib/apiservers/engine/backends/system.go (hunk 2 above)
Comment (sflxn) [inline lib/apiservers/engine/backends/system.go]: We do.  We need to clean up both errors and strings for both the personality and port layer server.  I don't want to do that with this PR or operation.  I have another GI to clean up strings in the personality server.

File: lib/apiservers/engine/backends/system.go (hunk 3 above)
Comment (sflxn) [inline lib/apiservers/engine/backends/system.go]: Yeah.

File: lib/apiservers/engine/backends/system.go (hunk 4 above)
Comment (sflxn) [inline lib/apiservers/engine/backends/system.go]: json-file appears to be the default Docker logging driver.

File: lib/apiservers/engine/backends/system.go (hunk 5 above)
Comment (sflxn) [inline lib/apiservers/engine/backends/system.go]: I was returning just the default docker info.  I'll move those to the "need to get from portlayer" section.

File: lib/apiservers/engine/backends/system.go (hunk 7 above)
Comment (sflxn) [inline lib/apiservers/engine/backends/system.go]: We could get the info ourselves, but this is the personality server.  Should it even know about any VMW specific APIs?

File: lib/apiservers/engine/backends/system.go (hunk 8 above)
Comment (sflxn) [inline lib/apiservers/engine/backends/system.go]: I thought we didn't want to leak information about the vsphere environment.

File: lib/apiservers/engine/backends/system.go (hunk 7 above)
Comment (hickeng) [inline lib/apiservers/engine/backends/system.go]: @sflxn It's not a vSphere API, it's a VIC package that's used for component configuration and status publishing.

File: lib/apiservers/engine/backends/container.go [hunk 9]
```diff
@@ -64,32 +63,32 @@ const (
 
 // ContainerExecCreate sets up an exec in a running container.
 func (c *Container) ContainerExecCreate(config *types.ExecConfig) (string, error) {
-	return "", fmt.Errorf("%s does not implement container.ContainerExecCreate", c.ProductName)
+	return "", fmt.Errorf("%s does not implement container.ContainerExecCreate", ProductName())
```
Comment (hickeng) [inline lib/apiservers/engine/backends/container.go]: @sflxn All of these should be pulled out into a message file - it shouldn't take long. The intent is that there is an actual github issue link included for unimplemented features.

File: lib/apiservers/engine/backends/system.go [hunk 10]
```diff
@@ -87,5 +198,73 @@ func (s *System) UnsubscribeFromEvents(chan interface{}) {
...
+	return "", "", fmt.Errorf("%s does not implement System.AuthenticateToRegistry", ProductName())
+}
+
+// Utility functions
+
+// Returns the name of the OS on the appliance vm.  Very likely Photon OS.  This
+// function takes the place of Docker's function that relies on other packages.
+// Since VIC will likely run on either Photon OS or devbox (ubuntu), we can
+// make some assumptions that avoid having to pull in other packages.
+func getOperatingSystem() string {
```
Comment (hickeng) [inline lib/apiservers/engine/backends/system.go]: This will do for now, but reporting the OS on the appliance is pointless in reality - it tells the user nothing about what they can actually run in containers. Luckily the bootstrap image and appliance are currently on the same version, but we cannot expect that to remain the case, so this will need to become significantly more involved in the future as it'll be reporting what the possible operating systems/variants are.

File: lib/apiservers/engine/backends/system.go [hunk 11]
```diff
@@ -15,31 +15,142 @@
...
+		//	IPv4Forwarding     bool
+		//	BridgeNfIptables   bool
+		//	BridgeNfIP6tables  bool `json:"BridgeNfIp6tables"`
+		HTTPProxy:  "N/A",
+		HTTPSProxy: "N/A",
+		NoProxy:    "N/A",
+	}
+
+	info.Plugins.Network = append(info.Plugins.Network, "N/A")
+	info.Plugins.Volume = append(info.Plugins.Volume, "N/A")
```
Comment (hickeng) [inline lib/apiservers/engine/backends/system.go]: I think this is the correct location to report the available volume locations that can be specified.

**Comment by hickeng** [comment]
@sflxn I'm okay with this going in as is for TP3 - I'd prefer it to have the volume locations (this is simply the keys from the VirtualContainerHostConfigSpec.VolumeLocations map) if possible.

File: lib/apiservers/engine/backends/container.go (hunk 9 above)
Comment (sflxn) [inline lib/apiservers/engine/backends/container.go]: #1099.  It won't take long, but I do have another higher priority issues to work on.  Also, if we're buying into the small commits regime, then we should have the strings be another PR.  It's on my list of items to work on immediately after our TP3 work.

**Comment by cgtexmex** [comment]
Help me to understand all the error message changes -- that's not required for info is it?  Is this change addressing two things?  Also...if the error messages aren't the way we want to go longer term do we really need this interim step?

**Comment by sflxn** [comment]
The error message thing does not address info.  I just wanted to centralize how we got productflags the way we do constants and globals in vicbackends to follow the same pattern we've been doing.  It started with trying to address comments about constants, which kinda balloon while I was waiting on our CI to work again.

**Comment by mdubya66** [comment]
lgtm

**Comment by hickeng** [comment]
LGTM
Viable placeholder data.

**Comment by matthewavery** [comment]
lgtm
